From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lovell Fuller <github@lovell.info>
Date: Wed, 19 Mar 2025 13:23:19 +0000
Subject: [PATCH 1/1] heifload: improve detection of seek beyond EOF

Upstream-Status: Submitted [https://github.com/libvips/libvips/pull/4424]

diff --git a/libvips/foreign/heifload.c b/libvips/foreign/heifload.c
index 1111111..2222222 100644
--- a/libvips/foreign/heifload.c
+++ b/libvips/foreign/heifload.c
@@ -1180,7 +1180,7 @@ vips_foreign_load_heif_wait_for_file_size(gint64 target_size, void *userdata)
 	result = vips_source_seek(heif->source, target_size, SEEK_SET);
 	vips_source_seek(heif->source, old_position, SEEK_SET);
 
-	if (result < 0)
+	if (result < 0 || old_position < 0)
 		/* Unable to seek to this point, so it's beyond EOF.
 		 */
 		status = heif_reader_grow_status_size_beyond_eof;
