From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 25 Jul 2024 11:25:00 +0200
Subject: [PATCH 1/1] Fix assertion failure in test suite


diff --git a/test/api/exceptions/unit-invalid.cpp b/test/api/exceptions/unit-invalid.cpp
index 1111111..2222222 100644
--- a/test/api/exceptions/unit-invalid.cpp
+++ b/test/api/exceptions/unit-invalid.cpp
@@ -16,8 +16,8 @@ TEST_CASE("invalid image", "[invalid]") {
                    Contains("Invalid or unsupported image format"));
     }
     SECTION("file") {
-        auto test_file = fixtures->dir + "/doesnotexist.jpg";
-        Status status = process_file(test_file, test_file, "");
+        auto test_file = fixtures->dir + "/looks-like-html.jpg";
+        Status status = process_file(test_file, nullptr, "");
 
         CHECK(!status.ok());
         CHECK(status.code() == static_cast<int>(Status::Code::InvalidImage));
diff --git a/test/api/fixtures/looks-like-html.jpg b/test/api/fixtures/looks-like-html.jpg
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/test/api/fixtures/looks-like-html.jpg
@@ -0,0 +1 @@
+<!DOCTYPE html>
